monitor/service.go

```diff
@@ -0,0 +1,306 @@
+// ExecuteStatement executes monitor-related query statements.
+func (s *Service) ExecuteStatement(stmt influxql.Statement) *influxql.Result {
+	switch stmt := stmt.(type) {
+	case *influxql.ShowStatsStatement:
+		return s.executeShowStatistics(stmt)
+	default:
+		panic(fmt.Sprintf("unsupported statement type: %T", stmt))
+	}
+}
+
+// executeShowStatistics returns the statistics of the registered monitor client in
+// the standard form expected by users of the InfluxDB system.
+func (s *Service) executeShowStatistics(q *influxql.ShowStatsStatement) *influxql.Result {
+	stats, _ := s.statistics()
+	rows := make([]*influxql.Row, len(stats))
+
+	for n, stat := range stats {
+		row := &influxql.Row{}
+		values := make([]interface{}, 0, len(stat.Tags)+len(stat.Values))
+
+		row.Columns = append(row.Columns, "name")
+		values = append(values, stat.Name)
+
+		for _, k := range stat.tagNames() {
+			row.Columns = append(row.Columns, k)
+			values = append(values, stat.Tags[k])
+		}
+		for _, k := range stat.valueNames() {
+			row.Columns = append(row.Columns, k)
+			values = append(values, stat.Values[k])
+		}
+		row.Values = [][]interface{}{values}
+		rows[n] = row
+	}
+	return &influxql.Result{Series: rows}
+}
+
+// statistics returns the combined statistics for all registered clients.
+func (s *Service) statistics() ([]*statistic, error) {
+	statistics := make([]*statistic, 0, len(s.registrations))
+	for _, r := range s.registrations {
+		stats, err := r.Client.Statistics()
+		if err != nil {
+			continue
+		}
+
+		// If a registered client has no field data, don't include it in the results
+		if len(stats) == 0 {
+			continue
+		}
+
+		statistics = append(statistics, newStatistic(r.name, r.tags, stats))
+	}
+	return statistics, nil
+}
+
+// storeStatistics writes the statistics to an InfluxDB system.
+func (s *Service) storeStatistics() error {
+	// XXX add tags such as local hostname and cluster ID
+	//a.Tags["clusterID"] = strconv.FormatUint(s.clusterID, 10)
+	//a.Tags["nodeID"] = strconv.FormatUint(s.nodeID, 10)
+	//a.Tags["hostname"] = s.hostname
+	return nil
+}
+
+// statistic represents the information returned by a single monitor client.
+type statistic struct {
+	Name   string
+	Tags   map[string]string
+	Values map[string]interface{}
+}
+
+// newStatistic returns a new statistic object. It ensures that tags are always non-nil.
+func newStatistic(name string, tags map[string]string, values map[string]interface{}) *statistic {
+	var a map[string]string
+
+	if tags == nil {
+		a = make(map[string]string)
+	} else {
+		a = tags
+	}
+
+	return &statistic{
+		Name:   name,
+		Tags:   a,
+		Values: values,
+	}
+}
+
+// tagNames returns a sorted list of the tag names, if any.
+func (s *statistic) tagNames() []string {
+	a := make([]string, 0, len(s.Tags))
+	for k, _ := range s.Tags {
+		a = append(a, k)
+	}
+	sort.Strings(a)
+	return a
+}
+
+// valueNames returns a sorted list of the value names, if any.
+func (s *statistic) valueNames() []string {
```

Comment:
Maybe this function is badly named. It is `valueNames` -- it the names of the values, what we normally call "Field Names".